WordPress/gutenberg
# Fluid Typography: Try adding a ceiling to the calculated minimum font size

## What?


A potential fix for https://github.com/WordPress/gutenberg/issues/49162

In the fluid typography output, explore setting a ceiling for the calculated minimum font size. This is to ensure that on smaller viewports, the font-size never gets too big for the viewport.

The PR currently uses a hard-coded `64px` as the ceiling (corresponding to `4rem`), which borrows from the minimum font size of the `xx-large` font size preset from TwentyTwentyThree theme. My thinking here, is that if it's a good default for the TT3 theme to use, then perhaps it's a good default for the custom fluid typography rules, too.

***Note:*** It is totally fine if the approach in this PR isn't viable — the intention is to poke at the existing fluid typography formula and learn more about what we need it to do.

## Why?


The current rules for custom font-sizes with `settings.typography.fluid` set to `true` in `theme.json` is for the maximum font-size to be multiplied by `0.75` to arrive at the minimum font-size. This works nicely for smaller headings, but the linear scale doesn't work well for really big font-sizes, as they'll still be too big in smaller viewports.

The idea in this PR is to see whether adding a ceiling to the minimum font size could resolve the issue for the most-part in the shorter-term. From some quick testing, it seems to work fairly well, with `clamp()` magic taking care of adjusting the font-size in between the minimum and maximum values.

Of course, this PR is also just a starting point for discussion — we can try out other ideas or go in a different direction with this.

## How?


* In both the JS and PHP for fluid typography, add a ceiling for minimum font size, hard-coded to `64px` / `4rem`.
* Use the ceiling for minimum font size only when a minimum font size has not been explicitly provided (so, we only invoke the ceiling for cases where the calculated rules apply, e.g. for custom font sizes)

## Testing Instructions





1. With a theme that uses `settings.typography.fluid` in `theme.json` (E.g. TwentyTwentyThree), add a bunch of paragraphs to a post or page using a variety of custom font sizes including really big ones like `8rem`. For a long paragraph, try using the text from the linked issue: `Loco brings you family recipes, fresh, simple, tasteful ingredients and a nice vibe.`
2. Publish the post and see how the fluid typography looks across a range of viewport sizes. Does the ceiling of `4rem` work okay? If there are a multiple large font sizes that are different at wide viewports (e.g. `8rem` next to `6rem`), do these scale down pleasingly?

For quicker testing, here is some test markup that includes a range of paragraphs:

<details>

<summary>Test markup of a variety of paragraph blocks at different custom font sizes</summary>

```html

<p style="font-size:8rem;line-height:1.1">Loco brings you family recipes, fresh, simple, tasteful ingredients and a nice vibe.</p>



<hr class="wp-block-separator has-alpha-channel-opacity"/>



<p style="font-size:8rem">A paragraph at 8rem</p>



<p style="font-size:6rem">A paragraph at 6rem</p>



<p style="font-size:4rem">A paragraph at 4rem</p>



<p style="font-size:3rem">A paragraph at 3rem</p>



<p style="font-size:2rem">A paragraph at 2rem</p>



<p style="font-size:1rem">A paragraph at 1rem</p>



<hr class="wp-block-separator has-alpha-channel-opacity"/>



<p style="font-size:128px">A paragraph at 128px</p>



<p style="font-size:96px">A paragraph at 96px</p>



<p style="font-size:64px">A paragraph at 64px</p>



<p style="font-size:48px">A paragraph at 48px</p>



<p style="font-size:32px">A paragraph at 32px</p>



<p style="font-size:16px">A paragraph at 16px</p>

```

</details>

### Testing Instructions for Keyboard


## Screenshots or screencast 

Using the test paragraph from the linked issue, the following shows the font-size at desktop screen size, and a 375px viewport:

| Desktop | Mobile viewport |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/226793903-a65ead61-187a-4246-860a-f3e6ff7bb542.png) | <img width="372" alt="image" src="https://user-images.githubusercontent.com/14988353/226793798-d5a166a0-bbf9-4553-86c4-89d0a53aee6a.png"> |

Screengrab:


https://user-images.githubusercontent.com/14988353/226794416-7ac764eb-4477-44d9-907e-3816add9e46e.mp4
